File: CTCI/linkedlists/rotatelist.js
function rotateList(list,n) {
    let root = list.head
    if (!root) return null 
    let fast = root 
    let slow = root 
    // find where fast starts 
    for (let i =0;i<n;i++) {
        fast = fast.next 
    }
    while (fast.next) {
        fast = fast.next 
        slow = slow.next 
    }
    let nextHead = slow.next
    slow.next = null 
    fast.next = root 
    list.head = nextHead
    return list
}

class List {
    constructor(){
        this.head = null 
    }
    add(data) {
        let node = new Node(data) 
        if (!this.head) this.head = node 
        else {
            let root = this.head 
            while (root.next) {
                root = root.next 
            }
            root.next = node 
        }
        return 
    }
    print() {
        let root = this.head 
        while (root) {
            console.log(root.data)
            root = root.next
        }
    }
}
class Node {
    constructor(data) {
        this.data = data
        this.next = null 
    }
   
}

let l = new List()
let node = new Node(1)
l.head = node 
l.add(2)
l.add(3)
l.add(4)
l.add(5)

console.log(rotateList(l,3).print()) // 3-4-5-1-2-null

File: CTCI/linkedlists/rotate_list.py
import list
def rotate_list(list,k):
    root = list.root
    if(not root):
        return None 
    fast = list.root
    slow = list.root 
    # fast should start at list length - k 
    for i in range(0,k):
        fast = fast.next 
    # the new tail is at list length - k 
    while (fast.next):
        fast = fast.next 
        slow = slow.next 
    # new head is slow.next value 
    new_head = slow.next
    # new tail is slow.next  
    slow.next = None
    # fast.next now points to old head 
    fast.next = root 
    # set new head 
    list.root = new_head 
    return list 

l = list.Linked_List()
for i in range(1,6):
    l.add(i)
def run():
    rotate_list(l,2).print() # 4 5 1 2 3 